substrate/primitives/core/src/crypto_bytes.rs

```diff
@@ -0,0 +1,233 @@
+	}
+}
+
+impl<const N: usize, T> PassBy for CryptoBytes<N, T> {
+	type PassBy = pass_by::Inner<Self, [u8; N]>;
+}
+
+impl<const N: usize, T> AsRef<[u8]> for CryptoBytes<N, T> {
+	fn as_ref(&self) -> &[u8] {
+		&self.0[..]
+	}
+}
+
+impl<const N: usize, T> AsMut<[u8]> for CryptoBytes<N, T> {
+	fn as_mut(&mut self) -> &mut [u8] {
+		&mut self.0[..]
+	}
+}
+
+impl<const N: usize, T> From<CryptoBytes<N, T>> for [u8; N] {
+	fn from(v: CryptoBytes<N, T>) -> [u8; N] {
+		v.0
+	}
+}
+
+impl<const N: usize, T> AsRef<[u8; N]> for CryptoBytes<N, T> {
+	fn as_ref(&self) -> &[u8; N] {
+		&self.0
+	}
+}
+
+impl<const N: usize, T> From<[u8; N]> for CryptoBytes<N, T> {
+	fn from(value: [u8; N]) -> Self {
+		Self::from_raw(value)
+	}
+}
+
+impl<const N: usize, T> TryFrom<&[u8]> for CryptoBytes<N, T> {
+	type Error = ();
+
+	fn try_from(data: &[u8]) -> Result<Self, Self::Error> {
+		if data.len() != N {
+			return Err(())
+		}
+		let mut r = [0u8; N];
+		r.copy_from_slice(data);
+		Ok(Self::from_raw(r))
+	}
+}
+
+impl<const N: usize, T> UncheckedFrom<[u8; N]> for CryptoBytes<N, T> {
+	fn unchecked_from(data: [u8; N]) -> Self {
+		Self::from_raw(data)
+	}
+}
+
+impl<const N: usize, T> core::ops::Deref for CryptoBytes<N, T> {
+	type Target = [u8];
+
+	fn deref(&self) -> &Self::Target {
+		&self.0
+	}
+}
+
+impl<const N: usize, T> CryptoBytes<N, T> {
+	/// Construct from raw array.
+	pub fn from_raw(inner: [u8; N]) -> Self {
+		Self(inner, PhantomData)
+	}
+
+	/// Construct from raw array.
+	pub fn to_raw(self) -> [u8; N] {
+		self.0
+	}
+
+	/// Return a slice filled with raw data.
+	pub fn as_array_ref(&self) -> &[u8; N] {
+		&self.0
+	}
+}
+
+impl<const N: usize, T> crate::ByteArray for CryptoBytes<N, T> {
+	const LEN: usize = N;
+}
+
+impl<const N: usize, T> FromEntropy for CryptoBytes<N, T> {
+	fn from_entropy(input: &mut impl codec::Input) -> Result<Self, codec::Error> {
+		let mut result = Self::default();
+		input.read(result.as_mut())?;
+		Ok(result)
+	}
+}
+
+impl<T> From<CryptoBytes<32, T>> for H256 {
+	fn from(x: CryptoBytes<32, T>) -> H256 {
+		H256::from(x.0)
+	}
+}
+
+impl<T> From<CryptoBytes<64, T>> for H512 {
+	fn from(x: CryptoBytes<64, T>) -> H512 {
+		H512::from(x.0)
+	}
+}
+
+impl<T> UncheckedFrom<H256> for CryptoBytes<32, T> {
+	fn unchecked_from(x: H256) -> Self {
+		Self::from_h256(x)
+	}
+}
+
+impl<T> CryptoBytes<32, T> {
+	/// A new instance from an H256.
+	pub fn from_h256(x: H256) -> Self {
+		Self::from_raw(x.into())
+	}
+}
+
+impl<T> CryptoBytes<64, T> {
+	/// A new instance from an H512.
+	pub fn from_h512(x: H512) -> Self {
+		Self::from_raw(x.into())
+	}
+}
+
+/// Tag used for generic public key bytes.
+pub struct PublicTag;
+
+/// Generic encoded public key.
+pub type PublicBytes<const N: usize, SubTag> = CryptoBytes<N, (PublicTag, SubTag)>;
+
+/// Tag used for generic signature bytes.
+pub struct SignatureTag;
+
+/// Generic encoded signature.
+pub type SignatureBytes<const N: usize, SubTag> = CryptoBytes<N, (SignatureTag, SubTag)>;
```

Comment:
nit: it is all mostly wrappers, but maybe you could add some tests? However I am not 100% they are actually necessary. Feel free to ignore.